src/Sparrow/Json/ArenaMemoryAllocator.cs

```diff
@@ -198,12 +189,23 @@ private void GrowArena(int requestedSize)
         private int GetPreferredSize(int requestedSize)
         {
             if (AvoidOverAllocation)
-                return Bits.NextPowerOf2(requestedSize);
+                return ApplyLimit(Bits.NextPowerOf2(requestedSize));
             
             // we need the next allocation to cover at least the next expansion (also doubling)
             // so we'll allocate 3 times as much as was requested, or as much as we already have
             // the idea is that a single allocation can server for multiple (increasing in size) calls
-            return Math.Max(Bits.NextPowerOf2(requestedSize) * 3, _initialSize);
+            return ApplyLimit(Math.Max(Bits.NextPowerOf2(requestedSize) * 3, _initialSize));
+
+            int ApplyLimit(int size)
+            {
+                if (SingleAllocationSizeLimit == null)
+                    return size;
+
+                if (SingleAllocationSizeLimit.Value > size)
+                    return Math.Max(Bits.NextPowerOf2(requestedSize), SingleAllocationSizeLimit.Value);
```

Comment:
we want to return SingleAllocationSizeLimit.Value here not `Math.Max` probably?

Reply:
Makes sense

Reply:
I'm wondering... maybe it is worth setting 'AvoidOverAllocation' to true on 32-bit?

Reply:
This isn't a global flag, it is used in specific circumstances, where we know we are done with writing and need the full operation

Reply:
full buffer, once.